src/lib/collab/provider.spec.js

```diff
@@ -0,0 +1,390 @@
+      const collab = new CollabProvider(config, service)
+      collab.initialize(getState)
+
+      const long = { version: 104, steps: [{}, {}, {}] } // start at 101
+      const medium = { version: 103, steps: [{}, {}] } // start at 101
+      const short = { version: 102, steps: [{}] } // start at 101
+      const first = { version: 101, steps: [{}] } // start at 100
+
+      collab.queueData(medium)
+      collab.queueData(short)
+      collab.queueData(first)
+      collab.queueData(long)
+
+      expect(collab.queue).toEqual([first, long, medium, short])
+    })
+  })
+
+  describe('catchup', () => {
+    const makeTheCatchupWait = (amount = 200, callback = () => {}) => {
+      channel.getSteps.mockImplementation(async version => {
+        await wait(amount)
+        return callback(version)
+      })
+    }
+
+    it('forbids to process new steps while catchup', async () => {
+      const nextVersion = version + steps.length
+      const getSteps = () => ({ steps, version: nextVersion })
+      makeTheCatchupWait(200, getSteps)
+
+      const collab = new CollabProvider(config, service)
+      collab.initialize(getState)
+
+      const catchup = collab.catchup()
+      collab.queueData({ version: nextVersion, steps: [{}, {}] })
+      collab.processQueue()
+      expect(collab.queue).not.toHaveLength(0)
+      await catchup
+    })
+
+    it('allows to process new steps after catchup', async () => {
+      const nextVersion = version + steps.length
+      const getSteps = () => ({ steps, version: nextVersion })
+      makeTheCatchupWait(200, getSteps)
+
+      const collab = new CollabProvider(config, service)
+      collab.initialize(getState)
+
+      const catchup = collab.catchup()
+      collab.queueData({ version: nextVersion + 2, steps: [{}, {}] })
+      collab.processQueue()
+
+      await catchup
+      expect(collab.queue).toHaveLength(0)
+    })
+
+    it('gets all missing steps from server', async () => {
+      const nextVersion = version + steps.length
+      const getSteps = () => ({ steps, version: nextVersion })
+      channel.getSteps.mockImplementation(getSteps)
+
+      const collab = new CollabProvider(config, service)
+      collab.initialize(getState)
+
+      const processSteps = jest.fn()
+      collab.on('data', processSteps)
+      await collab.catchup()
+
+      expect(processSteps).toHaveBeenCalled()
+      hasReceivedSteps(processSteps, steps, nextVersion)
+    })
+
+    it('reinit the full doc if the server cannot give the missing steps', async () => {
+      const getSteps = () => ({ doc, version: version })
+      channel.getSteps.mockImplementation(getSteps)
+
+      const collab = new CollabProvider(config, service)
+      collab.initialize(getState)
+
+      const init = jest.fn()
+      collab.on('init', init)
+      await collab.catchup()
+
+      expect(init).toHaveBeenCalled()
+      expect(init).toHaveBeenCalledWith({ doc, version, sid: sessionId })
```

Comment:
is it because you test the second call to init? 

If yes, I think that `expect(init).toHaveBeenNthCalledWith(2, { doc, version, sid: sessionId })` is easier to understand.

Reply:
The first check is useless, I am removing it.